### ISSUE TITLE
Go Concurrency Patterns: Pipelines and cancellation

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2713]
Go Concurrency Patterns: Pipelines and cancellation<br>
<br>
http://bit.ly/1za160V<br>
<br>
By Sameer Ajmani<br>
<br>
<div>
<div>
<p>13 March 2014</p>
<h4 id="TOC_1.">Introduction</h4>
<p>Go's concurrency primitives make it easy to construct streaming data pipelines that make efficient use of I/O and multiple CPUs. This article presents examples of such pipelines, highlights subtleties that arise when operations fail, and introduces techniques for dealing with failures cleanly.</p>
<h4 id="TOC_2.">What is a pipeline?</h4>
<p>There's no formal definition of a pipeline in Go; it's just one of many kinds of concurrent programs. Informally, a pipeline is a series of <i>stages</i> connected by channels, where each stage is a group of goroutines running the same function. In each stage, the goroutines</p>
<ul>
<li>receive values from <i>upstream</i> via <i>inbound</i> channels</li>
<li>perform some function on that data, usually producing new values</li>
<li>send values <i>downstream</i> via <i>outbound</i> channels</li>
</ul>
<p>Each stage has any number of inbound and outbound channels, except the first and last stages, which have only outbound or inbound channels, respectively. The first stage is sometimes called the <i>source</i> or <i>producer</i>; the last stage, the <i>sink</i> or <i>consumer</i>.</p>
<p>We'll begin with a simple example pipeline to explain the ideas and techniques. Later, we'll present a more realistic example.</p>
<h4 id="TOC_3.">Squaring numbers</h4>
<p>Consider a pipeline with three stages.</p>
<p>The first stage, <code>gen</code>, is a function that converts a list of integers to a channel that emits the integers in the list. The <code>gen</code> function starts a goroutine that sends the integers on the channel and closes the channel when all the values have been sent:</p>
<div>
<pre>
<span>func gen(nums ...int) &lt;-chan int {</span>
<span>    out := make(chan int)</span>
<span>    go func() {</span>
<span>        for _, n := range nums {</span>
<span>            out &lt;- n</span>
<span>        }</span>
<span>        close(out)</span>
<span>    }()</span>
<span>    return out</span>
<span>}</span>
</pre>
</div>
<p>The second stage, <code>sq</code>, receives integers from a channel and returns a channel that emits the square of each received integer. After the inbound channel is closed and this stage has sent all the values downstream, it closes the outbound channel:</p>
<div>
<pre>
<span>func sq(in &lt;-chan int) &lt;-chan int {</span>
<span>    out := make(chan int)</span>
<span>    go func() {</span>
<span>        for n := range in {</span>
<span>            out &lt;- n * n</span>
<span>        }</span>
<span>        close(out)</span>
<span>    }()</span>
<span>    return out</span>
<span>}</span>
</pre>
</div>
<p>The <code>main</code> function sets up the pipeline and runs the final stage: it receives values from the second stage and prints each one, until the channel is closed:</p>
<div>
<pre>
<span>func main() {</span>
<span>    // Set up the pipeline.</span>
<span>    c := gen(2, 3)</span>
<span>    out := sq(c)</span>

<span>    // Consume the output.</span>
<span>    fmt.Println(&lt;-out) // 4</span>
<span>    fmt.Println(&lt;-out) // 9</span>
<span>}</span>
</pre>
</div>
<p>Since <code>sq</code> has the same type for its inbound and outbound channels, we can compose it any number of times. We can also rewrite <code>main</code> as a range loop, like the other stages:</p>
<div>
<pre>
<span>func main() {</span>
<span>    // Set up the pipeline and consume the output.</span>
<span>    for n := range sq(sq(gen(2, 3))) {</span>
<span>        fmt.Println(n) // 16 then 81</span>
<span>    }</span>
<span>}</span>
</pre>
</div>
<h4 id="TOC_4.">Fan-out, fan-in</h4>
<p>Multiple functions can read from the same channel until that channel is closed; this is called <i>fan-out</i>. This provides a way to distribute work amongst a group of workers to parallelize CPU use and I/O.</p>
<p>A function can read from multiple inputs and proceed until all are closed by multiplexing the input channels onto a single channel that's closed when all the inputs are closed. This is called <i>fan-in</i>.</p>
<p>We can change our pipeline to run two instances of <code>sq</code>, each reading from the same input channel. We introduce a new function, <i>merge</i>, to fan in the results:</p>
<div>
<pre>
<span>func main() {</span>
<span>    in := gen(2, 3)</span>

<span>    // Distribute the sq work across two goroutines that both read from in.</span>
<span>    c1 := sq(in)</span>
<span>    c2 := sq(in)</span>

<span>    // Consume the merged output from c1 and c2.</span>
<span>    for n := range merge(c1, c2) {</span>
<span>        fmt.Println(n) // 4 then 9, or 9 then 4</span>
<span>    }</span>
<span>}</span>
</pre>
</div>
<p>The <code>merge</code> function converts a list of channels to a single channel by starting a goroutine for each inbound channel that copies the values to the sole outbound channel. Once all the <code>output</code> goroutines have been started, <code>merge</code> starts one more goroutine to close the outbound channel after all sends on that channel are done.</p>
<p>Sends on a closed channel panic, so it's important to ensure all sends are done before calling close. The <a href="https://golang.org/pkg/sync/#WaitGroup" target="_blank"><code>sync.WaitGroup</code></a> type provides a simple way to arrange this synchronization:</p>
<div>
<pre>
<span>func merge(cs ...&lt;-chan int) &lt;-chan int {</span>
<span>    <b>var wg sync.WaitGroup</b></span>
<span>    out := make(chan int)</span>

<span>    // Start an output goroutine for each input channel in cs.  output</span>
<span>    // copies values from c to out until c is closed, then calls wg.Done.</span>
<span>    output := func(c &lt;-chan int) {</span>
<span>        for n := range c {</span>
<span>            out &lt;- n</span>
<span>        }</span>
<span>        <b>wg.Done()</b></span>
<span>    }</span>
<span>    <b>wg.Add(len(cs))</b></span>
<span>    for _, c := range cs {</span>
<span>        go output(c)</span>
<span>    }</span>

<span>    // Start a goroutine to close out once all the output goroutines are</span>
<span>    // done.  This must start after the wg.Add call.</span>
<span>    go func() {</span>
<span>        <b>wg.Wait()</b></span>
<span>        close(out)</span>
<span>    }()</span>
<span>    return out</span>
<span>}</span>
</pre>
</div>
<h4 id="TOC_5.">Stopping short</h4>
<p>There is a pattern to our pipeline functions:</p>
<ul>
<li>stages close their outbound channels when all the send operations are done.</li>
<li>stages keep receiving values from inbound channels until those channels are closed.</li>
</ul>
<p>This pattern allows each receiving stage to be written as a <code>range</code> loop and ensures that all goroutines exit once all values have been successfully sent downstream.</p>
<p>But in real pipelines, stages don't always receive all the inbound values. Sometimes this is by design: the receiver may only need a subset of values to make progress. More often, a stage exits early because an inbound value represents an error in an earlier stage. In either case the receiver should not have to wait for the remaining values to arrive, and we want earlier stages to stop producing values that later stages don't need.</p>
<p>In our example pipeline, if a stage fails to consume all the inbound values, the goroutines attempting to send those values will block indefinitely:</p>
<div>
<pre>
<span>    // Consume the first value from the output.</span>
<span>    out := merge(c1, c2)</span>
<span>    fmt.Println(&lt;-out) // 4 or 9</span>
<span>    return</span>
<span>    // Since we didn't receive the second value from out,</span>
<span>    // one of the output goroutines is hung attempting to send it.</span>
<span>}</span>
</pre>
</div>
<p>This is a resource leak: goroutines consume memory and runtime resources, and heap references in goroutine stacks keep data from being garbage collected. Goroutines are not garbage collected; they must exit on their own.</p>
<p>We need to arrange for the upstream stages of our pipeline to exit even when the downstream stages fail to receive all the inbound values. One way to do this is to change the outbound channels to have a buffer. A buffer can hold a fixed number of values; send operations complete immediately if there's room in the buffer:</p>
<div>
<pre>
c := make(chan int, 2) // buffer size 2
c &lt;- 1  // succeeds immediately
c &lt;- 2  // succeeds immediately
c &lt;- 3  // blocks until another goroutine does &lt;-c and receives 1
</pre>
</div>
<p>When the number of values to be sent is known at channel creation time, a buffer can simplify the code. For example, we can rewrite <code>gen</code> to copy the list of integers into a buffered channel and avoid creating a new goroutine:</p>
<div>
<pre>
<span>func gen(nums ...int) &lt;-chan int {</span>
<span>    out := make(chan int, len(nums))</span>
<span>    for _, n := range nums {</span>
<span>        out &lt;- n</span>
<span>    }</span>
<span>    close(out)</span>
<span>    return out</span>
<span>}</span>
</pre>
</div>
<p>Returning to the blocked goroutines in our pipeline, we might consider adding a buffer to the outbound channel returned by <code>merge</code>:</p>
<div>
<pre>
<span>func merge(cs ...&lt;-chan int) &lt;-chan int {</span>
<span>    var wg sync.WaitGroup</span>
<span>    out := make(chan int, 1) // enough space for the unread inputs</span>
<span>    // ... the rest is unchanged ...</span>
</pre>
</div>
<p>While this fixes the blocked goroutine in this program, this is bad code. The choice of buffer size of 1 here depends on knowing the number of values <code>merge</code> will receive and the number of values downstream stages will consume. This is fragile: if we pass an additional value to <code>gen</code>, or if the downstream stage reads any fewer values, we will again have blocked goroutines.</p>
<p>Instead, we need to provide a way for downstream stages to indicate to the senders that they will stop accepting input.</p>
<h4 id="TOC_6.">Explicit cancellation</h4>
<p>When <code>main</code> decides to exit without receiving all the values from <code>out</code>, it must tell the goroutines in the upstream stages to abandon the values they're trying to send. It does so by sending values on a channel called <code>done</code>. It sends two values since there are potentially two blocked senders:</p>
<div>
<pre>
<span>func main() {</span>
<span>    in := gen(2, 3)</span>

<span>    // Distribute the sq work across two goroutines that both read from in.</span>
<span>    c1 := sq(in)</span>
<span>    c2 := sq(in)</span>

<span>    // Consume the first value from output.</span>
<span>    <b>done := make(chan struct{}, 2)</b></span>
<span>    out := merge(done, c1, c2)</span>
<span>    fmt.Println(&lt;-out) // 4 or 9</span>

<span>    // Tell the remaining senders we're leaving.</span>
<span>    <b>done &lt;- struct{}{}</b></span>
<span>    <b>done &lt;- struct{}{}</b></span>
<span>}</span>
</pre>
</div>
<p>The sending goroutines replace their send operation with a <code>select</code> statement that proceeds either when the send on <code>out</code> happens or when they receive a value from <code>done</code>. The value type of <code>done</code> is the empty struct because the value doesn't matter: it is the receive event that indicates the send on <code>out</code> should be abandoned. The <code>output</code> goroutines continue looping on their inbound channel, <code>c</code>, so the upstream stages are not blocked. (We'll discuss in a moment how to allow this loop to return early.)</p>
<div>
<pre>
<span>func merge(done &lt;-chan struct{}, cs ...&lt;-chan int) &lt;-chan int {</span>
<span>    var wg sync.WaitGroup</span>
<span>    out := make(chan int)</span>

<span>    // Start an output goroutine for each input channel in cs.  output</span>
<span>    // copies values from c to out until c is closed or it receives a value</span>
<span>    // from done, then output calls wg.Done.</span>
<span>    output := func(c &lt;-chan int) {</span>
<span>        for n := range c {</span>
<span>            select {</span>
<span>            case out &lt;- n:</span>
<span>            <b>case &lt;-done:</b></span>
<span>            }</span>
<span>        }</span>
<span>        wg.Done()</span>
<span>    }</span>
<span>    // ... the rest is unchanged ...</span>
</pre>
</div>
<p>This approach has a problem: <i>each</i> downstream receiver needs to know the number of potentially blocked upstream senders and arrange to signal those senders on early return. Keeping track of these counts is tedious and error-prone.</p>
<p>We need a way to tell an unknown and unbounded number of goroutines to stop sending their values downstream. In Go, we can do this by closing a channel, because <a href="https://golang.org/ref/spec#Receive_operator" target="_blank">a receive operation on a closed channel can always proceed immediately, yielding the element type's zero value.</a></p>
<p>This means that <code>main</code> can unblock all the senders simply by closing the <code>done</code> channel. This close is effectively a broadcast signal to the senders. We extend <i>each</i> of our pipeline functions to accept <code>done</code> as a parameter and arrange for the close to happen via a <code>defer</code> statement, so that all return paths from <code>main</code> will signal the pipeline stages to exit.</p>
<div>
<pre>
<span>func main() {</span>
<span>    // Set up a done channel that's shared by the whole pipeline,</span>
<span>    // and close that channel when this pipeline exits, as a signal</span>
<span>    // for all the goroutines we started to exit.</span>
<span>    <b>done := make(chan struct{})</b></span>
<span>    <b>defer close(done)</b></span>

<span>    in := gen(done, 2, 3)</span>

<span>    // Distribute the sq work across two goroutines that both read from in.</span>
<span>    c1 := sq(done, in)</span>
<span>    c2 := sq(done, in)</span>

<span>    // Consume the first value from output.</span>
<span>    out := merge(done, c1, c2)</span>
<span>    fmt.Println(&lt;-out) // 4 or 9</span>

<span>    <b>// done will be closed by the deferred call.</b></span>
<span>}</span>
</pre>
</div>
<p>Each of our pipeline stages is now free to return as soon as <code>done</code> is closed. The <code>output</code> routine in <code>merge</code> can return without draining its inbound channel, since it knows the upstream sender, <code>sq</code>, will stop attempting to send when <code>done</code> is closed. <code>output</code> ensures <code>wg.Done</code> is called on all return paths via a <code>defer</code> statement:</p>
<div>
<pre>
<span>func merge(done &lt;-chan struct{}, cs ...&lt;-chan int) &lt;-chan int {</span>
<span>    var wg sync.WaitGroup</span>
<span>    out := make(chan int)</span>

<span>    // Start an output goroutine for each input channel in cs.  output</span>
<span>    // copies values from c to out until c or done is closed, then calls</span>
<span>    // wg.Done.</span>
<span>    output := func(c &lt;-chan int) {</span>
<span>        <b>defer wg.Done()</b></span>
<span>        for n := range c {</span>
<span>            select {</span>
<span>            case out &lt;- n:</span>
<span>            case &lt;-done:</span>
<span>                <b>return</b></span>
<span>            }</span>
<span>        }</span>
<span>    }</span>
<span>    // ... the rest is unchanged ...</span>
</pre>
</div>
<p>Similarly, <code>sq</code> can return as soon as <code>done</code> is closed. <code>sq</code> ensures its <code>out</code> channel is closed on all return paths via a <code>defer</code> statement:</p>
<div>
<pre>
<span>func sq(done &lt;-chan struct{}, in &lt;-chan int) &lt;-chan int {</span>
<span>    out := make(chan int)</span>
<span>    go func() {</span>
<span>        <b>defer close(out)</b></span>
<span>        for n := range in {</span>
<span>            select {</span>
<span>            case out &lt;- n * n:</span>
<span>            case &lt;-done:</span>
<span>                <b>return</b></span>
<span>            }</span>
<span>        }</span>
<span>    }()</span>
<span>    return out</span>
<span>}</span>
</pre>
</div>
<p>Here are the guidelines for pipeline construction:</p>
<ul>
<li>stages close their outbound channels when all the send operations are done.</li>
<li>stages keep receiving values from inbound channels until those channels are closed or the senders are unblocked.</li>
</ul>
<p>Pipelines unblock senders either by ensuring there's enough buffer for all the values that are sent or by explicitly signalling senders when the receiver may abandon the channel.</p>
<h4 id="TOC_7.">Digesting a tree</h4>
<p>Let's consider a more realistic pipeline.</p>
<p>MD5 is a message-digest algorithm that's useful as a file checksum. The command line utility <code>md5sum</code> prints digest values for a list of files.</p>
<div>
<pre>
% md5sum *.go
d47c2bbc28298ca9befdfbc5d3aa4e65  bounded.go
ee869afd31f83cbb2d10ee81b2b831dc  parallel.go
b88175e65fdcbc01ac08aaf1fd9b5e96  serial.go
</pre>
</div>
<p>Our example program is like <code>md5sum</code> but instead takes a single directory as an argument and prints the digest values for each regular file under that directory, sorted by path name.</p>
<div>
<pre>
% go run serial.go .
d47c2bbc28298ca9befdfbc5d3aa4e65  bounded.go
ee869afd31f83cbb2d10ee81b2b831dc  parallel.go
b88175e65fdcbc01ac08aaf1fd9b5e96  serial.go
</pre>
</div>
<p>The main function of our program invokes a helper function <code>MD5All</code>, which returns a map from path name to digest value, then sorts and prints the results:</p>
<div>
<pre>
<span>func main() {</span>
<span>    // Calculate the MD5 sum of all files under the specified directory,</span>
<span>    // then print the results sorted by path name.</span>
<span>    <b>m, err := MD5All(os.Args[1])</b></span>
<span>    if err != nil {</span>
<span>        fmt.Println(err)</span>
<span>        return</span>
<span>    }</span>
<span>    var paths []string</span>
<span>    for path := range m {</span>
<span>        paths = append(paths, path)</span>
<span>    }</span>
<span>    <b>sort.Strings(paths)</b></span>
<span>    for _, path := range paths {</span>
<span>        fmt.Printf("%x  %s\n", m[path], path)</span>
<span>    }</span>
<span>}</span>
</pre>
</div>
<p>The <code>MD5All</code> function is the focus of our discussion. In <a href="https://blog.golang.org/pipelines/serial.go" target="_self">serial.go</a>, the implementation uses no concurrency and simply reads and sums each file as it walks the tree.</p>
<div>
<pre>
<span>// MD5All reads all the files in the file tree rooted at root and returns a map</span>
<span>// from file path to the MD5 sum of the file's contents.  If the directory walk</span>
<span>// fails or any read operation fails, MD5All returns an error.</span>
<span>func MD5All(root string) (map[string][md5.Size]byte, error) {</span>
<span>    m := make(map[string][md5.Size]byte)</span>
<span>    <b>err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {</b></span>
<span>        if err != nil {</span>
<span>            return err</span>
<span>        }</span>
<span>        if !info.Mode().IsRegular() {</span>
<span>            return nil</span>
<span>        }</span>
<span>        <b>data, err := ioutil.ReadFile(path)</b></span>
<span>        if err != nil {</span>
<span>            return err</span>
<span>        }</span>
<span>        <b>m[path] = md5.Sum(data)</b></span>
<span>        return nil</span>
<span>    })</span>
<span>    if err != nil {</span>
<span>        return nil, err</span>
<span>    }</span>
<span>    return m, nil</span>
<span>}</span>
</pre>
</div>
<h4 id="TOC_8.">Parallel digestion</h4>
<p>In <a href="https://blog.golang.org/pipelines/parallel.go" target="_self">parallel.go</a>, we split <code>MD5All</code> into a two-stage pipeline. The first stage, <code>sumFiles</code>, walks the tree, digests each file in a new goroutine, and sends the results on a channel with value type <code>result</code>:</p>
<div>
<pre>
<span>type result struct {</span>
<span>    path string</span>
<span>    sum  [md5.Size]byte</span>
<span>    err  error</span>
<span>}</span>
</pre>
</div>
<p><code>sumFiles</code> returns two channels: one for the <code>results</code> and another for the error returned by <code>filepath.Walk</code>. The walk function starts a new goroutine to process each regular file, then checks <code>done</code>. If <code>done</code> is closed, the walk stops immediately:</p>
<div>
<pre>
<span>func sumFiles(done &lt;-chan struct{}, root string) (&lt;-chan result, &lt;-chan error) {</span>
<span>    // For each regular file, start a goroutine that sums the file and sends</span>
<span>    // the result on c.  Send the result of the walk on errc.</span>
<span>    c := make(chan result)</span>
<span>    errc := make(chan error, 1)</span>
<span>    <b>go func() {</b></span>
<span>        var wg sync.WaitGroup</span>
<span>        err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {</span>
<span>            if err != nil {</span>
<span>                return err</span>
<span>            }</span>
<span>            if !info.Mode().IsRegular() {</span>
<span>                return nil</span>
<span>            }</span>
<span>            wg.Add(1)</span>
<span>            <b>go func() {</b></span>
<span>                data, err := ioutil.ReadFile(path)</span>
<span>                select {</span>
<span>                <b>case c &lt;- result{path, md5.Sum(data), err}:</b></span>
<span>                <b>case &lt;-done:</b></span>
<span>                }</span>
<span>                wg.Done()</span>
<span>            }()</span>
<span>            // Abort the walk if done is closed.</span>
<span>            select {</span>
<span>            <b>case &lt;-done:</b></span>
<span>                return errors.New("walk canceled")</span>
<span>            default:</span>
<span>                return nil</span>
<span>            }</span>
<span>        })</span>
<span>        // Walk has returned, so all calls to wg.Add are done.  Start a</span>
<span>        // goroutine to close c once all the sends are done.</span>
<span>        <b>go func() {</b></span>
<span>            wg.Wait()</span>
<span>            <b>close(c)</b></span>
<span>        }()</span>
<span>        // No select needed here, since errc is buffered.</span>
<span>        <b>errc &lt;- err</b></span>
<span>    }()</span>
<span>    return c, errc</span>
<span>}</span>
</pre>
</div>
<p><code>MD5All</code> receives the digest values from <code>c</code>. <code>MD5All</code> returns early on error, closing <code>done</code> via a <code>defer</code>:</p>
<div>
<pre>
<span>func MD5All(root string) (map[string][md5.Size]byte, error) {</span>
<span>    // MD5All closes the done channel when it returns; it may do so before</span>
<span>    // receiving all the values from c and errc.</span>
<span>    <b>done := make(chan struct{})</b></span>
<span>    <b>defer close(done)</b></span>

<span>    <b>c, errc := sumFiles(done, root)</b></span>

<span>    m := make(map[string][md5.Size]byte)</span>
<span>    for r := range c {</span>
<span>        if r.err != nil {</span>
<span>            return nil, r.err</span>
<span>        }</span>
<span>        m[r.path] = r.sum</span>
<span>    }</span>
<span>    if err := &lt;-errc; err != nil {</span>
<span>        return nil, err</span>
<span>    }</span>
<span>    return m, nil</span>
<span>}</span>
</pre>
</div>
<h4 id="TOC_9.">Bounded parallelism</h4>
<p>The <code>MD5All</code> implementation in <a href="https://blog.golang.org/pipelines/parallel.go" target="_self">parallel.go</a> starts a new goroutine for each file. In a directory with many large files, this may allocate more memory than is available on the machine.</p>
<p>We can limit these allocations by bounding the number of files read in parallel. In <a href="https://blog.golang.org/pipelines/bounded.go" target="_self">bounded.go</a>, we do this by creating a fixed number of goroutines for reading files. Our pipeline now has three stages: walk the tree, read and digest the files, and collect the digests.</p>
<p>The first stage, <code>walkFiles</code>, emits the paths of regular files in the tree:</p>
<div>
<pre>
<span>func walkFiles(done &lt;-chan struct{}, root string) (&lt;-chan string, &lt;-chan error) {</span>
<span>    paths := make(chan string)</span>
<span>    errc := make(chan error, 1)</span>
<span>    <b>go func() {</b></span>
<span>        // Close the paths channel after Walk returns.</span>
<span>        <b>defer close(paths)</b></span>
<span>        // No select needed for this send, since errc is buffered.</span>
<span>        <b>errc &lt;- filepath.Walk(root, func(path string, info os.FileInfo, err error) error {</b></span>
<span>            if err != nil {</span>
<span>                return err</span>
<span>            }</span>
<span>            if !info.Mode().IsRegular() {</span>
<span>                return nil</span>
<span>            }</span>
<span>            select {</span>
<span>            <b>case paths &lt;- path:</b></span>
<span>            <b>case &lt;-done:</b></span>
<span>                return errors.New("walk canceled")</span>
<span>            }</span>
<span>            return nil</span>
<span>        })</span>
<span>    }()</span>
<span>    return paths, errc</span>
<span>}</span>
</pre>
</div>
<p>The middle stage starts a fixed number of <code>digester</code> goroutines that receive file names from <code>paths</code> and send <code>results</code> on channel <code>c</code>:</p>
<div>
<pre>
<span>func digester(done &lt;-chan struct{}, paths &lt;-chan string, c chan&lt;- result) {</span>
<span>    <b>for path := range paths {</b></span>
<span>        data, err := ioutil.ReadFile(path)</span>
<span>        select {</span>
<span>        case c &lt;- result{path, md5.Sum(data), err}:</span>
<span>        case &lt;-done:</span>
<span>            return</span>
<span>        }</span>
<span>    }</span>
<span>}</span>
</pre>
</div>
<p>Unlike our previous examples, <code>digester</code> does not close its output channel, as multiple goroutines are sending on a shared channel. Instead, code in <code>MD5All</code> arranges for the channel to be closed when all the <code>digesters</code> are done:</p>
<div>
<pre>
<span>    // Start a fixed number of goroutines to read and digest files.</span>
<span>    <b>c := make(chan result)</b></span>
<span>    var wg sync.WaitGroup</span>
<span>    const numDigesters = 20</span>
<span>    wg.Add(numDigesters)</span>
<span>    for i := 0; i &lt; numDigesters; i++ {</span>
<span>        go func() {</span>
<span>            <b>digester(done, paths, c)</b></span>
<span>            wg.Done()</span>
<span>        }()</span>
<span>    }</span>
<span>    go func() {</span>
<span>        wg.Wait()</span>
<span>        <b>close(c)</b></span>
<span>    }()</span>
</pre>
</div>
<p>We could instead have each digester create and return its own output channel, but then we would need additional goroutines to fan-in the results.</p>
<p>The final stage receives all the <code>results</code> from <code>c</code> then checks the error from <code>errc</code>. This check cannot happen any earlier, since before this point, <code>walkFiles</code> may block sending values downstream:</p>
<div>
<pre>
<span>    m := make(map[string][md5.Size]byte)</span>
<span>    for r := range c {</span>
<span>        if r.err != nil {</span>
<span>            return nil, r.err</span>
<span>        }</span>
<span>        m[r.path] = r.sum</span>
<span>    }</span>
<span>    // Check whether the Walk failed.</span>
<span>    <b>if err := &lt;-errc; err != nil {</b></span>
<span>        return nil, err</span>
<span>    }</span>
<span>    return m, nil</span>
<span>}</span>
</pre>
</div>
<h4 id="TOC_10.">Conclusion</h4>
<p>This article has presented techniques for constructing streaming data pipelines in Go. Dealing with failures in such pipelines is tricky, since each stage in the pipeline may block attempting to send values downstream, and the downstream stages may no longer care about the incoming data. We showed how closing a channel can broadcast a "done" signal to all the goroutines started by a pipeline and defined guidelines for constructing pipelines correctly.</p>
<p>Further reading:</p>
<p>By Sameer Ajmani</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via blog.golang.org https://blog.golang.org<br>
<br>
February 14, 2019 at 02:27PM
